design/observerability-2.md

```diff
@@ -0,0 +1,196 @@
+> is to overcome the limitation of JAX-RS specification in that it does not require
+> implementations to call a ContainerResponseFilter to finish exchanges that led
+> to an unmapped exception. That's why we can't use a ContainerResponseFilter to
+> 'finish' the tracking of a REST call and update the corresponding metrics.
+
+it is necessary to properly track all non-successful responses.
+
+The last part is tracking external IdP responses. Implementing this part requires changing the existing code around
+`SocialIdentityProvider`s and `OIDCIdentityProvider`. All `SocialIdentityProvider` implementations usually throw
+`IdentityBrokerException`s. The exception needs to be enhanced to contain an information about that provider has thrown
+it. Then, the implementation should involve updating all places in `AbstractOAuth2IdentityProvider` and add
+proper checks in all `catch` clauses. Once an `IdentityBrokerException` is caught, and it contains a provider ID,
+it should result in updating certain metrics in the metrics components after logging an error.
+
+## Logging architecture
+
+Collecting logs in an OpenShift cluster is a cross-cutting concern that needs a unified solution. OpenShift provides a
+Cluster Logging Operator, that represents a typical EFK stack implementation:
+
+![EFK Stack](img/observability-efk.png)
+
+The Cluster Logging Operator uses Kibana (Open Distro) to view logs. The log-collecting stack requires annotating a
+namespace with `openshift.io/cluster-monitoring: "true"` label.
+
+#### Alerts on log errors
+
+Unfortunately, the Cluster Logging Operator (and Kibana in this case) does not allow to set any alerts
+(Watch UI is part of the "x-pack" distribution) nor to plot the output in Grafana.
+This is very likely to change in the near future. Both AOS Logging and Managed Applications
+teams are likely to introduce [Loki](https://grafana.com/oss/loki/) - a Grafana add-on for viewing
+logs. So, before implementing this piece of functionality - please consult it with MAS and AOS Logging Teams.
+
+Depending on timing and other teams progress, it is possible to implement error rate tracking using metrics. This
+however, introduces a few design flaws:
+- A new `LoggerAdapter` would need to be created. Whenever an error is logged in Keycloak's codebase, the `LoggerAdapter`
+will increase the error log counter.
+- All `Logger#getLogger` calls would need to be replaced to acquire the `LoggerAdapter` instead of pure
+JBoss Logging `Logger`.
+- The error-rate metric will miss all error emitted by Wildfly (as they will be thrown in different subsystems).
+
+Keycloak.x distribution offers logging the number of log events by enabling `quarkus.log.metrics.enabled`. The functionality
+has been implemented in Quarkus using [this Pull Request](https://github.com/quarkusio/quarkus/pull/11224).
+
+Having the above in mind, if the implementation choice is to expose the number error in logs using metrics endpoint,
```

Comment:
Is it possible to expose metrics for the underlying causes of the errors?

Reply:
With the logging improvements I mentioned in the new version of the design - it should be possible to do that. Loki should enable you to select logs with a specific phrase, like "KC-SERVICES0002". If we used the prefix consistently, you should be able to build such metrics ad-hoc.